src/actions/user-actions.js

```diff
@@ -0,0 +1,84 @@
+import userService from '../services/user-service';
+import {REGISTER_USER, FIND_ALL_USERS, FIND_USER_BY_ID, LOGIN_USER, LOGOUT_USER} from "./user-constants";
+
+export const registerUser = (dispatch, user) => {
+    // retrieve all users
+    userService
+        .findAllUsers()
+        .then(actualUsers => {
+            let flag = 0
+            // of the actual users, iterate through each and check if the username has not been taken.
+            actualUsers.forEach(actualUser => {
+                if(actualUser.username === user.username){
+                    // if taken, mark it as such.
+                    flag = -1
+                }
+            })
+            // if the username has not been taken, register the new user
+            if(flag !== -1){
+                userService.registerUser(user).then(response => dispatch({
+                    type: REGISTER_USER,
+                    user: response
+                }))
+            }
+            else{
+                alert("Username has already been taken.")
+            }
+        })
+};
+
+export const loginUser = (dispatch, user) => {
+    // retrieve all users
+    userService
+        .findAllUsers()
```

Comment:
This is not needed you can just call the login like below.  The login endpoint checks if the user is in the database. I would attempt to hit the endpoint with a user that does not exist and see what the response is. If it returns an error please use `catch()` to handle or if it returns something different use that logic to handle it. I would recommend fixing this before merging into main. Thanks and please let me know if you have any questions.

```
userService.login(user).then(response => dispatch({...
```